### PR TITLE
POC: Implement filtering data work

### PR DESCRIPTION
**Pull request by kaywux**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -383,3 +383,5 @@ require (
 	github.com/ziutek/mymysql v1.5.4 // indirect
 	golang.org/x/exp v0.0.0-20230321023759-10a507213a29
 )
+
+replace go.viam.com/api => /Users/katherinewu/viam/api
```
Comment (kaywux) [inline go.mod:387]: API changes: https://github.com/viamrobotics/api/pull/316

File: filterDemo/katcam/katcam.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+func (fs filterStream) Next(ctx context.Context) (image.Image, func(), error) {
+	if ctx.Value(data.CtxKeyDM) != true {
+		return nil, nil, errors.New("Cannot access filter stream if not DM collector")
+	}
+	// return fs.cameraStream.Next(ctx)
+
+	// // Case 1: Only store one out of every 10 captures.
+	// *fs.count++
```
Comment (tessavitabile) [inline filterDemo/katcam/katcam.go:158]: Needs to be thread-safe

File: filterDemo/katcam/katcam.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	// // Case 1: Only store one out of every 10 captures.
+	// count *int32
+
+	// Case 2:
+	visionService vision.Service
+}
+
+func (fs filterStream) Next(ctx context.Context) (image.Image, func(), error) {
+	if ctx.Value(data.CtxKeyDM) != true {
+		return nil, nil, errors.New("Cannot access filter stream if not DM collector")
```
Comment (tessavitabile) [inline filterDemo/katcam/katcam.go:153]: If it's not a data management collector, we want to return the underlying camera stream.

File: data/collector.go [hunk 4]
```diff
@@ -186,7 +193,12 @@ func (c *collector) getAndPushNextReading() {
 	reading, err := c.captureFunc(c.cancelCtx, c.params)
 	timeReceived := timestamppb.New(c.clock.Now().UTC())
 	if err != nil {
-		c.captureErrors <- errors.Wrap(err, "error while capturing data")
+		if errors.Is(err, ErrNoCaptureToStore) {
+			// Capture nothing because of filter module, do not store.
+			c.logger.Infoln("Handling empty capture data...")
```
[the remark this answers is not in the document]
Comment (erh) [reply data/collector.go:198]: How would this work with a python module?

[the remark this answers is not in the document]
Comment (kaywux) [reply data/collector.go:198]: In their python module the user will raise an Exception with the same string message as `ErrNoCaptureToStore`, which the collector will [catch](https://github.com/viamrobotics/rdk/pull/2712/files/6d6276e886a3e8bf3e174efcbe5a607a647216a9#diff-eb904023d3acc2651764502ee0cfde50ac7d1d7439a7d23a08db874eae79dffbR90) and throw ErrNoCaptureToStore - which will then be caught here.

We can note in the documentation that the error returned in the module must have the same message as what we're expecting for data capture to work as expected.

[the remark this answers is not in the document]
Comment (erh) [reply data/collector.go:198]: That's not done yet though right?

**Pull request closed without merging**